src/particles/distribution/Waterbag.H

```diff
@@ -0,0 +1,125 @@
+/* Copyright 2021 Ji Qiang, Axel Huebl
+ *
+ * This file is part of ImpactX.
+ *
+ * License: BSD-3-Clause-LBNL
+ */
+#ifndef IMPACTX_DISTRIBUTION_WATERBAG
+#define IMPACTX_DISTRIBUTION_WATERBAG
+
+#include <AMReX_Random.H>
+#include <AMReX_REAL.H>
+
+#include <cmath>
+
+
+namespace impactx
+{
+namespace distribution
+{
+    struct Waterbag
+    {
+        /** A Waterbag distribution
+         *
+         * Return sampling from a 6D Waterbag distribution.
+         *
+         * @param sigx,sigy,sigt for zero correlation, these are the related
+         *                                   RMS sizes (length); ...
+         * @param sigpx,sigpy,sigpt RMS momentum
+         * @param muxpx,muypy,mutpt correlation length-momentum
+         */
+         Waterbag(amrex::ParticleReal const sigx, amrex::ParticleReal const sigy,
+         amrex::ParticleReal const sigt,amrex::ParticleReal const sigpx,
+         amrex::ParticleReal const sigpy,amrex::ParticleReal const sigpt,
+         amrex::ParticleReal const muxpx=0.0, amrex::ParticleReal const muypy=0.0,
+         amrex::ParticleReal const mutpt=0.0
+         )
+          : m_sigmaX(sigx),m_sigmaY(sigy),m_sigmaT(sigt),m_sigmaPx(sigpx),m_sigmaPy(sigpy),
+            m_sigmaPt(sigpt),m_muxpx(muxpx),m_muypy(muypy),m_mutpt(mutpt)
+          {
+          }
+
+        /** Return 1 6D particle coordinate
+         *
+         * @param x particle position in x
+         * @param y particle position in y
+         * @param t particle position in t
+         * @param px particle momentum in x
+         * @param py particle momentum in y
+         * @param pt particle momentum in t
+         * @param engine a random number engine (with associated state)
+         */
+        AMREX_GPU_HOST_DEVICE AMREX_FORCE_INLINE
+        void operator() (
+                amrex::ParticleReal & x,
+                amrex::ParticleReal & y,
+                amrex::ParticleReal & t,
+                amrex::ParticleReal & px,
+                amrex::ParticleReal & py,
+                amrex::ParticleReal & pt,
+                amrex::RandomEngine const& engine) const {
+
+            using namespace amrex::literals;
+
+            amrex::ParticleReal ln1,norm,u1,u2;
+            amrex::ParticleReal g1,g2,g3,g4,g5,g6;
+            amrex::ParticleReal root,a1,a2;
+
+            constexpr amrex::ParticleReal pi = 3.14159265358979_prt;
```

Comment:
@cemitch99 suggested on Slack to add a few inline comments of the form
```C++
// now we calculate this... then that
```

here, to make the logic here easier to follow.

Reply:
@cemitch99 @qianglbl please add this as needed in a follow-up PR :)